feat(forms): `ngForm` element selector has been deprecated

alan-agius4 opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
`ngForm` element selector has been deprecated in favor of `ng-form`

Align with styleguides

 Now deprecated:
 ```
 <ngForm #myForm="ngForm">
 ```

 After:
 ```
 <ng-form #myForm="ngForm">
 ```

Closes: #23678


## What is the new behavior?


## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

kara commented on packages/forms/src/directives/ng_form.ts:
```diff
@@ -55,16 +55,28 @@ const resolvedPromise = Promise.resolve(null);
  * refrain from using the `formGroup` directive.
  *
+ * Support for using `ngForm` element selector has been deprecated in Angular v6 and will be removed
+ * in Angular v7.
```
Generally, we try not to remove within 2 versions of the next major. So this should be v9.

Also can you add more about how this change is occurring to keep this selector consistent with other core Angular selectors?

kara reviewed (changes requested):
Generally when we deprecate something, we try to have an extended migration plan to make it easier for users to transition to the new API.  So we can't really deprecate this selector without the rest of the plan in place.

In this case, we'd probably want to display a console warning if the `ngForm` selector was used (probably by separating that out into its own directive) and some affordance (probably a provider) that allows the warning to be silenced.

You can see https://github.com/angular/angular/pull/22633 as an example of a console warning as part of a deprecation plan.  Can you try to implement something like that?

alan-agius4 commented:
I quite like your suggestions and take on implementing the warning messages that can be suppressed. And I do agree that the best way to go most likely is to have a separate directive as I won’t like to to use the reflector or elementRef to get the selector used.

I can implement something similar. I just need a couple of mdays though as I am currently rather busy.

alan-agius4 commented:
It shouldn't as `@ViewChild(NgForm)` will still point to the old directive

trotyl commented:
Understood now, new directive only for warning purpose?

alan-agius4 commented:
@trotyl indeed. I actually managed to find some time to do it today. I'll push an update soon, which you can see.

alan-agius4 commented:
@kara, PR updated

@trotyl, check here: https://github.com/angular/angular/pull/23721/files#diff-09f29ed452087bcb0965b730a9dad964R224

alan-agius4 commented:
@kara bump.

alan-agius4 commented:
@kara any update on this please?

alan-agius4 commented:
@kara / @IgorMinar can we get this merged?, it's been quite a while.

kara commented on packages/forms/src/directives/ng_form_selector_warning.ts:
```diff
@@ -0,0 +1,46 @@
+export const NG_FORM_SELECTOR_WARNING = new InjectionToken('NgFormSelectorWarning');
+
+/**
+ * This directive is soly used to display warnings when the deprecated `ngForm` selector is used.
```
Typo: soly -> solely

kara commented on packages/forms/src/directives/ng_form_selector_warning.ts:
```diff
@@ -0,0 +1,46 @@
+import {Directive, Inject, InjectionToken, Optional} from '@angular/core';
+
+/**
+ * Token to provide turn off warning when using 'ngForm' deprecated selector.
```
> Token to provide to turn off the warning when using 'ngForm' deprecated selector.

kara commented on packages/forms/src/directives/ng_form_selector_warning.ts:
```diff
@@ -0,0 +1,46 @@
+@Directive({selector: 'ngForm'})
+export class NgFormSelectorWarning {
+  /**
+   * Static property used to track FormControlDirective. Used to support warning config of "once".
```
Copy-paste error? This warning shouldn't occur on `FormControlDirective`.  Maybe

> Static property used to track whether the deprecation warning for this selector has been sent. Used to support warning config of "once".

alan-agius4 replied:
oops yeah

kara commented on packages/forms/src/directives/ng_form_selector_warning.ts:
```diff
@@ -0,0 +1,46 @@
+   */
+  static _ngFormWarning = false;
+
+  constructor(@Optional() @Inject(NG_FORM_SELECTOR_WARNING) private ngFormWarning: string|null) {
```
I don't think we need to save `ngFormWarning`, since we are only using it in this function.

kara commented on packages/forms/src/directives/ng_form_selector_warning.ts:
```diff
@@ -0,0 +1,46 @@
+  constructor(@Optional() @Inject(NG_FORM_SELECTOR_WARNING) private ngFormWarning: string|null) {
+    if (!ngFormWarning || (ngFormWarning === 'once' && !NgFormSelectorWarning._ngFormWarning) ||
+        ngFormWarning === 'always') {
+      console.warn(`
```
Template-driven form error messages are stored in `template_driven_errors.ts`. Can you move this `console.warn` into a function there? Here you'd have something like:

```ts
TemplateDrivenErrors.ngFormWarning();

```

kara commented on packages/forms/src/directives/ng_form_selector_warning.ts:
```diff
@@ -0,0 +1,46 @@
+      console.warn(`
+      It looks like you're using 'ngForm'.
+
+      Support for using 'ngForm' element selector has been deprecated in Angular v6 and will be removed
```
Missing a `the`

>Support for using the 'ngForm' element selector

kara commented on packages/forms/src/directives/ng_form_selector_warning.ts:
```diff
@@ -0,0 +1,46 @@
+      Support for using 'ngForm' element selector has been deprecated in Angular v6 and will be removed
+      in Angular v9.
+
+      Use 'ng-form' instead.
```
I think this needs some example code for clarity. e.g.

Before:
```html
<ngForm #myForm="ngForm">
```

After:
```
<ng-form #myForm="ngForm">
```

kara commented on packages/forms/test/template_integration_spec.ts:
```diff
@@ -1630,6 +1630,61 @@ import {NgModelCustomComp, NgModelCustomWrapper} from './value_accessor_integrat
+              .toMatch(/It looks like you're using 'ngForm'/gi);
+        });
+
+        it(`should warn for each instance by default`, () => {
```
We should probably warn just once by default (don't want it to be too overwhelming). Then if people want to track down all instances of `ngForm` to remove them, they can switch to `always`.  This would also be more consistent with our other deprecation warnings, which default to once.

alan-agius4 replied:
Yes, maybe due to large apps, it's better to use `once` by default

kara reviewed (changes requested):
Sorry for the delay in reviewing! Thanks for taking the time to implement a full solution. Some comments below

alan-agius4 committed:
refactor(forms): `ngForm` element selector has been deprecated in favor of `ng-form`

This has been deprecated to keep selector consistent with other core Angular selectors.  As element selectors are in kebab-case.

 Now deprecated:
 ```
 <ngForm #myForm="ngForm">
 ```

 After:
 ```
 <ng-form #myForm="ngForm">
 ```

You can also choose to supress this warnings by providing a config for `FormsModule` during import:

```ts
imports: [
 FormsModule.withConfig({warnOnDeprecatedNgFormSelector: 'never'});
]

Closes: #23678

alan-agius4 commented:
@kara, thanks for reviewing this, and apologies for the typos. 
I have pushed a new version with the requested changes.

kitkat96 commented:
Alan,
First of all let me tell you that I have no idea what I'm doing and I'm.
Just trying to keep trackers away. My stuff had been getting changed in
time log and erased . I started with GitHub hoping I would understand and I
just don't no don't even know where to run anything

On Wed, Aug 1, 2018, 12:13 AM Alan Agius <notifications@github.com> wrote:

> *@alan-agius4* commented on this pull request.
> ------------------------------
>
> In packages/forms/src/directives/ng_form_selector_warning.ts
> <https://github.com/angular/angular/pull/23721#discussion_r206756778>:
>
> > +
> +/**
> + * Token to provide turn off warning when using 'ngForm' deprecated selector.
> + */
> +export const NG_FORM_SELECTOR_WARNING = new InjectionToken('NgFormSelectorWarning');
> +
> +/**
> + * This directive is soly used to display warnings when the deprecated `ngForm` selector is used.
> + *
> + * @deprecated in Angular v6 and will be removed in Angular v9.
> + *
> + */
> +@Directive({selector: 'ngForm'})
> +export class NgFormSelectorWarning {
> +  /**
> +   * Static property used to track FormControlDirective. Used to support warning config of "once".
>
> oops yeah
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/23721#discussion_r206756778>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/An13vEIuIeoDJ1ODzCX8fx6-3LhjZq0-ks5uMTkJgaJpZM4TzntJ>
> .
>

kara reviewed (approved):
LGTM! Thanks for the quick turnaround.

kara commented:
[presubmit](http://test/OCL:207025798:BASE:207026260:1533170948157:74d7ac06)

alan-agius4 commented:
@kara looks like one of your labels has been misplaced.

kara commented:
@alan-agius4 Was waiting for the internal Google presubmit to go through (that's usually what causes a delay at this stage)

alan-agius4 commented:
I meant that one of your labels ended up as a comment rather than a label.
The ‘presubmit’ has not been added as a label.



On Thu, 02 Aug 2018 at 06:11, Kara <notifications@github.com> wrote:

> @alan-agius4 <https://github.com/alan-agius4> Was waiting for the
> internal Google presubmit to go through (that's usually what causes a delay
> at this stage)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/angular/angular/pull/23721#issuecomment-409799707>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AQv-WkV69ory-lYAuV9l2o8VZsZmPJMdks5uMnvfgaJpZM4TzntJ>
> .
>

kara commented:
@alan-agius4 That's intentionally a comment. It's a link to the presubmit results so the Angular team caretaker can see the test results before merging into master.

kara commented:
No worries! I think only a few of us actually do it.
